pcdsdevices/zoomtelescope.py

```diff
@@ -0,0 +1,95 @@
+import logging
+
+from ophyd import Device, EpicsSignal, EpicsSignalRO, Component as Cpt
+
+logger = logging.getLogger(__name__)
+
+
+class ZoomTelescope(Device):
+    """
+    Zoom telescope
+
+    Device for the zoom telescope used in the MODS TILEs.
+
+    Parameters
+    ----------
+    prefix : str
+        Base prefix of the zoom telescope
+
+    name : str
+        Name of zoom telescope object
+
+    Examples
+    --------
+
+    zt = ZoomTelescope('LAS:ZOOM:01', name='LAS Zoom 1')
+
+    zt.zoom = 1.5 # set the zoom to 1.5x magnification
+
+    zt.tweak_val = 0.05 # set tweak parameter to 0.05
+
+    zt.tweak_plus()     # tweak the zoom by +zt.tweak_val
+
+    zt.tweak_minus()     # tweak the zoom by +zt.tweak_val
+    """
+
+    # Calibration parameters
+    # Not sure about putting these calibration parameters in here, since they
+    # should never really be required during normal operation, but will leave
+    # them in for debugging during commissioning.
+    _l2_p0 = Cpt(EpicsSignalRO, ':CAL:L2:P0', kind='omitted')
+    _l2_p1 = Cpt(EpicsSignalRO, ':CAL:L2:P1', kind='omitted')
+
+    _l3_p0 = Cpt(EpicsSignalRO, ':CAL:L3:P0', kind='omitted')
+    _l3_p1 = Cpt(EpicsSignalRO, ':CAL:L3:P1', kind='omitted')
+    _l3_p2 = Cpt(EpicsSignalRO, ':CAL:L3:P2', kind='omitted')
+    _l3_p3 = Cpt(EpicsSignalRO, ':CAL:L3:P3', kind='omitted')
+    _l3_p4 = Cpt(EpicsSignalRO, ':CAL:L3:P4', kind='omitted')
+    _l3_p5 = Cpt(EpicsSignalRO, ':CAL:L3:P5', kind='omitted')
+
+    _min_zoom = Cpt(EpicsSignalRO, ':CAL:MIN_ZOOM', kind='config')
+    _max_zoom = Cpt(EpicsSignalRO, ':CAL:MAX_ZOOM', kind='config')
+
+    # Control parameters
+    _req_zoom = Cpt(EpicsSignal, ':REQ_ZOOM', kind='normal')
+
+    _tweak_plus = Cpt(EpicsSignal, ':TWEAK_ZOOM_PLUS.PROC', kind='normal')
+    _tweak_minus = Cpt(EpicsSignal, ':TWEAK_ZOOM_MINUS.PROC', kind='normal')
+    _tweak_val = Cpt(EpicsSignal, ':TWEAK', kind='normal')
+
+    _permission = Cpt(EpicsSignalRO, ':PERMISSION_STATUS', kind='normal')
+
+    @property
+    def tweak_val(self):
+        """The amount to tweak the magnification when calling tweak_plus or
+        tweak_minus."""
+        tweak = self._tweak_val.get()
```

Comment:
We need to be careful of doing EPICS gets in attribute access, this can cause various problems because getattr is expected to be short/instant.

Reply:
So the alternative would be skip the properties and just use <component>.get()/.put(<val>)? 

I like using properties for this because I get tired of adding .put()/.get() at the end of everything (especially interactively), but if it's problematic then I can switch to this method.

Reply:
I'm going to merge this with the getattr/setattr as long as you're sure it's what you want with these devices. So take what I'm saying here (and in most other comments) as a suggestion rather than a mandate.

I agree that the getattr/setattr is convenient. It becomes confusing though as some things are getattr/setattr and some things are not. Does the user need to memorize which devices have this set up?

A long getattr causes the most frequent issues with tab completion in ipython. This is usually why the tab will be slow. Perhaps there's no reason to care about the slow tabs when your device is disconnected though.

Reply:
Properties are _really_ convenient - there's no arguing there. I was a big proponent of using them everywhere in ophyd and devices in general, but have slowly shifted my opinion over time.

Consider the following:
* Properties hide `args` and `kwargs` that might get passed to `.get` and `.set` (or `.put`)
* Properties may raise `TimeoutError` making `device.tweak_val` take a few seconds in places where you wouldn't expect it to
* Properties don't get mapped into the bluesky data model (that is, `.read()` and `.describe()` won't know about them)
* Properties won't return a status object indicating completion in the background (as `.set()` will)

The solution for most of these is "get used to calling `.get`/`.set`", but also to consider adding components (making them visible to bluesky) with alternative signal classes such as `ophyd.signal.DerivedSignal` or `ophyd.signal.AttributeSignal`. Unfortunately, documentation of these is nonexistent outside of the source code. ☹️

Reply:
Thank you both for your sage advice. Consider me a convert. :) I'll make this change. 

I will take a look at DerivedSignal and Attribute signal as well. Thanks!